### ISSUE TITLE
Question: why use Cmd.run over Cmd.action?

### ISSUE THREAD

**Issue by heygrady** [redux-loop/redux-loop#159]
I'm new here :) Just trying to get a handle on how this would work.

I'm confused why `Cmd.run` is the preferred example. While it's nice to have some automated handling of promises, there's already middleware for all of those things. It *feels* like `Cmd.run` is subverting the typical dispatch flow of a redux application.

On the other hand, using `Cmd.action` *feels* more like classic redux to me. 

#### The [docs](https://github.com/redux-loop/redux-loop#write-a-reducer-with-some-cmds-side-effects) suggest using `Cmd.run`

The func you call with `Cmd.run` isn't an action and isn't dispatched. This is *maybe* a little confusing. It appears that redux-loop expects your func to return a promise.

```js
// ...
function reducer(state = initialState, action) {
  switch(action.type) {
  case 'INIT':
    return loop(
      {...state, initStarted: true},
      // why is this better?
      Cmd.run(fetchUser, { // <-- not an action, does not get dispatched, skips middleware entirely
        successActionCreator: userFetchSuccessfulAction, // <-- nice for promises, but what about sagas?
        failActionCreator: userFetchFailedAction,
        args: ['123']
      })
    );
  // ...
  default:
    return state;
  }
}
```

#### Using `Cmd.action` seems to make more sense (to me)

Why not rely on good-ole dispatching? I can see the utility of "catching" actions in your reducer and re-dispatching them. This may indeed be the missing piece of the redux puzzle. In the example below, imagine we are implementing `fetchUser` as a thunk, or a promise middleware, or a saga. Under-the-hood we may be "fetching" that user from any number of async locations, like `fetch` or localstorage.

```js
// ...
function reducer(state = initialState, action) {
  switch(action.type) {
  case 'INIT':
    return loop(
      {...state, initStarted: true},
      Cmd.action(fetchUser(123)) // <-- let the middleware sort it out
    );
  // ...
  default:
    return state;
  }
}
```

#### Why dispatching actions may be better

Our fetching lifecycle could be more complex than simply success/fail. Like, what if the "fetch" pulls in data from three different sources? We'd want a success, fail, retry for each source before the final item is ready. These complex actions are why complex middleware was created.

The very existence of the `Cmd.dispatch` and `Cmd.getState` symbols is fishy to me.

```js
import { Cmd } from 'redux-loop';

// maybe I already have a thunk
const fetchUserThunk = (payload) => (dispatch, getState) => {
  dispatch(fetchUserStarted(payload));
  // imagine some useful logic
  dispatch(fetchUserSuccess(payload));
  dispatch(fetchUserFail(payload));
}

// let's hack a redux-loop-thunk surrogate
Cmd.thunk(func => Cmd.run(func, { args: [Cmd.dispatch, Cmd.getState] }));

Cmd.thunk(fetchUserThunk(123)); // <-- a thunk that (perhaps unexpectedly) skips middleware entirely
Cmd.action(fetchUserThunk(123)); // <-- a thunk that uses middleware as expected
```

Having used redux-saga extensively in the past, I can see that `Cmd.run` might be a better way to initiate a saga, rather than initializing a bunch of "watchers" to gobble actions. But the docs don't really make that use-case clear (to me).

Given the myriad ways people are capturing actions and feeding them to functions using middleware, it seems like lower friction to rely on `Cmd.action`. The bonus is that other parts of your application can listen for your actions and deal with them as appropriate. The downside to using `Cmd.run` is that the rest of the app is shielded from what's happening.

I'm not so well-versed in Elm so I may be missing the point entirely. It seems like redux-loop is handling the case that certain types of actions are *related* to a specific reducer, but those actions should actually kick-off a complex set of lifecycle actions. It makes sense to allow for dispatching actions from reducers for this case, but managing complex lifecycle actions is already well-handled by existing middleware.

I'm just leaving this note to hopefully help people who are confused, like me :)

**Comment by bdwain** [comment]
Hi @heygrady. Thanks for the feedback. I understand why you'd lean towards Cmd.action being more familiar with thunk and sagas. I haven't used redux-loop much in the way you are describing. It sounds like it could lead to interesting integrations with other libraries. I'd be happy to hear your thoughts on it if you end up exploring it more. 

Personally, I think of Redux Loop as more of an alternative to thunk and sagas than a complement to them. You may be able to use them together, but it could get more complex with little extra benefit over choosing just one. Redux loop solves some of the issues that come from using thunks and sagas, though like every solution, has both pros and cons. 

For your main question though, I can try to go through the reasoning that generally leads towards using Cmd.run in that example over Cmd.action.

First, if you're not familiar with it, I encourage you to read about [the elm architecture](https://guide.elm-lang.org/architecture). Redux is [partially based on it](http://redux.js.org/docs/introduction/PriorArt.html), and Redux-Loop tries to fill in the missing pieces and turn redux into a full port of the elm architecture.

One of the core ideas of the elm architecture and redux-loop is that the reducer (or updater in elm) is responsible for managing the control flow of the application. Not just updating the state, but determining what happens _next_ (aka what side effects to run). 

The main benefit of this is that you're not handling the same actions in multiple places. Your reducers control the entire flow of your application, instead of splitting it up among reducers and thunks/sagas. If you are making a request to an API, in one spot you can say "set state.fetching to true, call fetchItem, and dispatch this result when you're done". I've found this to be very useful when trying to trace through a code path in my application.

Many people feel this violates the single responsibility principle because your reducers are now updating state and dealing with side effects. My thought is that there is no reason to distinguish between the two as far as responsibility goes. They are both affecting the control flow of the application. Another way of thinking of this is that the state update is just one of many effects that the action can have on the application. 

Generally, I agree with the notion that actions should represent a thing that happened, rather than a thing you want to happen. This can simplify your action log a lot and make it more clear why things are happening, not just what is happening (See [this stack overflow post too](https://stackoverflow.com/questions/35406707/do-events-and-actions-have-a-11-relationship-in-redux/35410524#35410524)). That's why I generally don't use Cmd.action all that often. It has its uses in certain scenarios, but typically if an action is telling me that FOO happened and I want to run BAR side effect because of that, I'd rather specify that I am running BAR in the same spot i do the state update, because another action won't give me any more info about what has happened.

> Our fetching lifecycle could be more complex than simply success/fail. Like, what if the "fetch" pulls in data from three different sources? We'd want a success, fail, retry for each source before the final item is ready. These complex actions are why complex middleware was created.

It definitely could. However, a single function will always have a single result, and nothing prevents you from composing side effects into a single more complex side effect the same way you compose them into a single thunk. In fact, you could even replicate thunks fully by using Cmd.getState and Cmd.dispatch also. Though that brings up your last point. 

> The very existence of the Cmd.dispatch and Cmd.getState symbols is fishy to me.

I agree. It's one of those things that I don't like in theory but it just ends up being the most practical solution in some cases. I probably could document better that it is there more as an escape hatch (like refs in react if you are familiar) than to be a standard thing to use. See #67 for my original motivation for this.

**Comment by heygrady** [comment]
Thanks for the detailed response. This really helped. I think I'm understanding it a little bit better.

When I was working with sagas on a previous project that used react-redux-starter-kit I ended up creating some helper functions ([redux-saga-watch-actions](https://github.com/heygrady/redux-saga-watch-actions)) for making a saga behave more like a reducer. I was originally pretty enthused about how redux-loop looked because of how similar it appeared.

Looking at what you wrote above, it seems like `Cmd.run` is sort of a drop-in replacement for `sagaMiddleware.run`.

```js
import { Cmd } from 'redux-loop'
import createSagaMiddleware from 'redux-saga'

const sagaMiddleware = createSagaMiddleware() // <-- be sure to add this to your redux middleware
const runSaga = saga => sagaMiddleware.run(saga)

// let's hack in a saga-aware run
Cmd.runSaga = (saga, action) => Cmd.run(() => runSaga(saga, action))

// here's a simple sags
const fetchUser = function * (action) {
   try {
      const user = yield call(Api.fetchUser, action.payload.userId);
      yield put({type: "USER_FETCH_SUCCEEDED", user: user});
   } catch (e) {
      yield put({type: "USER_FETCH_FAILED", message: e.message});
   }
}

// now we can launch a saga to manage our updater's side-effects
Cmd.runSaga(fetchUser)
```

**Comment by bdwain** [comment]
I'm not totally sure I see the relationship between Cmd.run and sagamiddleware.run. Don't you just say sagaMiddleware.run once to set up all of your sagas? I thought they were more like a separate reducer tree (but sagas not reducers) that you initialize once when your app starts (like as part of the process of configuring the store). 

Cmd.run represents a single run of a single function. It's just a description of a single side effect you want to happen. It seems to be more related to the call saga effect, which seems to be similar.

**Comment by heygrady** [comment]
Right. That makes total sense. Although `sagaMiddleware.run` is essentially the same as the `call` effect as well.

The typical way they promote using sagas is exactly as I've implemented in [redux-saga-watch-action](https://heygrady.github.io/redux-saga-watch-actions/). Basically, redux-saga provides a [`takeEvery`](https://redux-saga.js.org/docs/api/index.html#takeeverypattern-saga-args) effect which, when you set it to run forever, will capture every matching action.

But that's not the only way you can run a saga. You can also just run it once. The ability to `take` an action is the quirky bit that allows it to be like a reducer, except a saga. However, "running" a saga is just running a generator function that yields effects; `take` is just one effect. The most useful part of redux-saga is a full-featured effects runner. It makes generators less cumbersome and transforms them into something quite elegant. It's a great way to write async code.

But, like you were alluding to, the practical way to set redux-saga up is as a "watcher" that kicks off a specific generator function when a specific action is dispatched. This is why redux-loop was initially intriguing to me; because I had done some work making sagas work like reducers and thought that the way redux-loop glued them together was pretty elegant.

In particular, redux-loop seems to have a better story for how you initialize your "watchers." They are *part* of the reducer/updater, so they "happen" whenever a specific action reaches that reducer. That's pretty slick!

This is just a long way of saying that I don't think it's necessary to consider it to be redux-saga versus redux-loop. Redux-saga doesn't have a very good story for how I initialize my side effects function. And redux-loop is not nearly as fully featured an effects runner as redux-saga.

#### A note on  `mergeChildReducers`
I also think you might consider releasing `mergeChildReducers` as a standalone package. It feels similar to [`reduceReducers`](https://github.com/acdlite/reduce-reducers) and immediately solved a need I had when trying to construct a reducer tree using [`handleActions`](https://github.com/reduxactions/redux-actions/blob/master/docs/api/handleAction.md#handleactions). There are times when `combineReducers` is too restrictive. Sometimes you need the parent key to catch actions as well, which `combineReducers` doesn't allow.

In short, `mergeChildReducers` is useful technology that I could make use of with or without redux-loop.

**Comment by bdwain** [comment]
I see. So you're thinking about something like Cmd.saga to run arbitrary sagas. Something like `Cmd.saga(someSaga, {args?, otherStuff...})`?

Were you thinking of this for more complex side effects? like the one you described earlier.
> Like, what if the "fetch" pulls in data from three different sources? We'd want a success, fail, retry for each source before the final item is ready. These complex actions are why complex middleware was created.

It might be possible to do that. Are you thinking the main benefit would be that the sagas are declarative and so easier to test? Because you can already run arbitrarily complex side effects like this

```js
Cmd.run(complexEffectWithManyTasks, {args: [Cmd.getState, Cmd.dispatch]})
```

and then basically do whatever you want inside that effect. However, I can see sagas being easier to test than a side effect like that, which is basically just a thunk.

That's a really interesting idea that I never thought about. I'd be interested in seeing how it worked. I don't even think we'd need to depend on sagas directly because you could just optionally pass a runSaga method in when creating the store. 

Really I think the ideal solution is if there were a common declarative effects running package, and then you could choose how you decide to run effects (Cmds vs takeEvery and similar things) separately from how you run them. But considering how many people use sagas so far, I think it would only make sense if they split their effects stuff into a separate package and we went with that. I don't think people would want to migrate to a new effect runner.

Are you interested in creating a PR to try this? If not I can, but it was your idea so if you're interested go for it. 

As for mergeChildReducers, thanks I'm glad you find it useful. I suppose I could release it as a standalone package. We'd still need the redux loop one because it has loop specific stuff in it, but I guess it would be useful. It feels weird to release a function that's just a simplified version of combineReducers, but I guess reduceReducers is useful too so why not. Thanks for the idea.

Also I forgot to mention this before, but Cmd.run does not require your function returns a promise. It detects if the return value is a promise or not and from that decides whether the value it uses is the promise resolution/rejection or the return value/thrown value.

**Comment by bdwain** [comment]
@ashishsc @willisplummer I'd also be interested in hearing your opinions on this.

**Comment by heygrady** [comment]
Thanks for the offer. I'm currently swamped on a few work things so I wouldn't be able to make progress on it right away. Feel free to run with it, and thanks for being so graciously considering my feedback.

#### Redux-saga as an effects runner
That's all it is already, except they have some boilerplate configuration, similar to your `install`, to get access to the store. They're allowing you to read from the store with `select` and dispatch with `put`. So, they're just wrapping `dispatch` and `getState` the same as redux-loop is.

#### Using redux-saga within redux-loop as a plugin effects runner
If the redux-loop install method has access to the store, it can initialize redux-saga completely behind the scenes. Or, if someone already had redux-saga configured, they could provide their `runSaga` function to the `install` function to enable `Cmd.runSaga`.

You only need redux-saga configured as middleware if you intend to `take`. Of course, you need `dispatch` and `getState` to have a useful saga. But those aren't exactly dependent on redux-saga intercepting actions in redux's middleware. Even so, if you were writing a saga and intending to use it for redux-loop effects you probably wouldn't have much use for `take`.

#### Returning promises, managing effects
While redux-loop doesn't *require* promises, it's kind of neat that redux-saga actually returns one from the `runSaga` function. Redux-saga returns a ["task"](https://github.com/redux-saga/redux-saga/tree/master/docs/api#task), which contains a promise (and some other useful stuff).

```js
const sagaMiddleware = createSagaMiddleware()
const runSaga = saga => sagaMiddleware.run(saga)

runSaga(saga, action).done // <-- redux-saga returns a "task" which contains a promise
```

By allowing the plugin effects runner to return a promise, redux-loop can keep its baseline effects engine (which is just a success/fail promise AFAIK) while supporting a wide range of more complex engines. In the case of redux-saga, you can detect if the previous saga has completed, or even cancel it!

#### Standardizing tasks, supporting pluggable effects runners
The basics of the redux-saga tasks are pretty easy to understand and covers all your bases. All you really need is something like the [flux standard action standard](https://github.com/acdlite/flux-standard-action), but for tasks.

- baseline, return a promise
- fully featured, return a task object
  - The [task object](https://github.com/redux-saga/redux-saga/tree/master/docs/api#task) in the redux-saga api looks like a good start for that definition
- allow for custom `runSaga` or `runThunk` or `runPromise` to be passed in as part of `install`
  - require that these functions return a promise/task to get the full functionality
  - full functionality is when redux-loop is able to manage your task for you, like handling options for auto-cancelling and other fancy features.

**Comment by bdwain** [comment]
Cool I'll play around with this. I don't think we'd necessarily want to automatically pull in saga, at least at first. Maybe down the road. To start we'd probably rely on passing a runsaga method to the store creator

**Comment by gniquil** [comment]
@heygrady @bdwain This is an interesting discussion, and I have some hard learned lesson about these 2 libs perhaps I could share. At work, we started using Saga first, coupled with typescript. We eventually got so fed up and moved to loop for the following reason:

1. Fragmentation of your state and effect. As @bdwain alluded earlier, we had a complete separate folder of sagas, independent of the reducers (and some thunks). This made many of our business logic difficult to track down as it is all over the place. Often we can see that something is triggered by an action, which leads to some state changes, and then in sagas folder some more delayed/async things that will eventually lead to additional state changes. This was a major source of frustrution

2. Where is exactly your state? One of the more serious problem I had with saga is sagas implicitly hold state that is not visible anywhere. One of the coolest example of saga is the auth/login/logout flow (i think it was demoed in many videos and blog posts). In Rich Hickey's easy vs simple terms, I believe this is easy but definitely not simple. In that flow, you can feel that somewhere in that `yield take(login)` than `yield take(logout)`, we are implicitly saying that the state of the application is "logged in", but this is not marked anywhere. This is why with many of our sagas, we have "put(someAction)" everywhere essentially behaves like setters to "notify" our state that our it is changed. I believe this a manifestation of fundamental problem of implicit state with saga. In fact, at work, we ban using `Cmd.getState` and `Cmd.dispatch` as effect arguments to prevent creating effects that are too encompassing and that it's actually holding implicit state. 

3. Modularity and typing. Since we use typescript, typing sagas is very problematic. What does getState return? The root state. But if you want to be modular, work on a sub component at a time, where do you get such a root state? You can't. Essentially everything is global. In fact, this is one of the biggest problem with redux as well. Actually not with redux itself, but with `combineReducers`. It fosters this global-ness of working that makes decomposition impossible (sagas is very much this as well). Also, think about the signature of your reducer and sagas. What is that action? If you use combineReducers or sagas, it's the application action. That means every sub reducer/saga needs to be aware of all actions handled by all other reducers/sagas. Redux-loop, on the other hand, follows the original Elm approach, with its nice `Cmd.map` function. This allows you to truly decompose your reducer/action into smaller chunks. You can reasonable your app one small chunk at a time. With typescript, we actually purposely leave out the `default` case in all reducers as the `Action` is well defined (not global), so discriminated union type in typescript understands we handled all actions. And when we add an action, typescript actually gives us error saying we have unhandled action. (So challenge to you is try not to use combinedReducers.) Check out Richard Feldman's talk on scaling Elm applications. In particular, if you do adopt the Elm approach, you could ask yourself why reducers have to return just the state, or just a loop ([State, Cmd]), why not return more things? After all, aren't they just normal functions used by other reducers which are also just normal functions. The only reducer that needs to conform to that special return type signature is the root reducer. Bottom line, redux-loop is much more modular than saga or the way redux is used by most people. And because everything is just function composition, you have all the leverage you have with functional programming. And it's type safe as well. 

I hope this helps.
